File: app/dtos/health_check.py
from datetime import date, datetime
from typing import Annotated

from pydantic import BaseModel, Field

from app.dtos.base import BaseSerializerModel
from app.models.health_check import AppGroup, CkdStage, DialysisType, UrineResult


class HealthCheckCreateRequest(BaseModel):
    checked_date: Annotated[date, Field(description="검진일 (YYYY-MM-DD)")]

    # 혈압 (mmHg)
    systolic_bp: Annotated[int, Field(ge=60, le=300, description="수축기혈압 (mmHg)")]
    diastolic_bp: Annotated[int, Field(ge=40, le=200, description="이완기혈압 (mmHg)")]

    # 혈액 검사
    fasting_glucose: Annotated[float, Field(ge=50.0, le=700.0, description="공복혈당 (mg/dL)")]
    creatinine: Annotated[
        float | None,
        Field(None, ge=0.1, le=30.0, description="혈청 크레아티닌 (mg/dL) — eGFR 계산에 사용"),
    ]
    total_cholesterol: Annotated[float | None, Field(None, ge=50.0, le=700.0, description="총 콜레스테롤 (mg/dL)")]
    hdl_cholesterol: Annotated[float | None, Field(None, ge=10.0, le=200.0, description="HDL 콜레스테롤 (mg/dL)")]
    triglycerides: Annotated[float | None, Field(None, ge=20.0, le=2000.0, description="중성지방 (mg/dL)")]
    ldl_cholesterol: Annotated[
        float | None,
        Field(None, ge=10.0, le=500.0, description="LDL 콜레스테롤 (mg/dL) — 입력값 우선, 미입력 시 Friedewald 계산"),
    ]
    hemoglobin: Annotated[float | None, Field(None, ge=3.0, le=25.0, description="헤모글로빈 (g/dL)")]
    ast: Annotated[float | None, Field(None, ge=0.0, le=2000.0, description="AST (U/L)")]
    alt: Annotated[float | None, Field(None, ge=0.0, le=2000.0, description="ALT (U/L)")]
    urine_protein: Annotated[UrineResult | None, Field(None, description="요단백 (양성/음성)")]
    urine_glucose: Annotated[UrineResult | None, Field(None, description="요당 (양성/음성)")]

    # 신체 측정
    weight: Annotated[float, Field(ge=20.0, le=300.0, description="체중 (kg)")]
    height: Annotated[float, Field(ge=100.0, le=250.0, description="신장 (cm)")]
    waist_circumference: Annotated[float | None, Field(None, ge=40.0, le=200.0, description="허리둘레 (cm)")]
    # dialysis_type 입력 제거 — 최신 문진(LifestyleSurvey)에서 조회·미러링(단일 진실)


class HealthCheckResponse(BaseSerializerModel):
    id: int
    user_id: int
    checked_date: date

    # 혈압
    systolic_bp: int
    diastolic_bp: int

    # 혈액 검사
    fasting_glucose: float
    creatinine: float | None
    total_cholesterol: float | None
    hdl_cholesterol: float | None
    triglycerides: float | None
    ldl_cholesterol: float | None = None
    hemoglobin: float | None = None
    ast: float | None = None
    alt: float | None = None
    urine_protein: UrineResult | None = None
    urine_glucose: UrineResult | None = None

    # 신체 측정
    weight: float
    height: float
    bmi: float
    waist_circumference: float | None

    # AI / CKD-EPI 예측 결과 (비동기 처리, 처음엔 null)
    egfr_estimated: float | None
    ckd_risk_score: float | None
    ckd_stage: CkdStage | None
    app_group: AppGroup | None
    dialysis_type: DialysisType | None = None

    # 세이프티 가드 메시지 (위험 수치 감지 시)
    safety_warning: str | None = None

    created_at: datetime


class HealthCheckListResponse(BaseSerializerModel):
    total: int
    items: list[HealthCheckResponse]


# ── SHAP 리포트 DTO ──────────────────────────────────────────────────────────


class ShapItem(BaseModel):
    feature: str
    value: float
    shap: float
    note: str | None = None  # 모델1만 note 사용, 모델2 items는 None
    # 임상 단계 라벨 (예: "양성", "위험", "주의", "정상" …) — classify_shap_items 분류 기준.
    status: str | None = None
    # 임상 위험 수준 색상 코드 (good/info/caution/warnLight/danger) — 프론트 배경색 렌더용.
    status_level: str | None = None
    # 모델2 전용 — m2_in_normal 게이트 분류 결과 ("improve" | "maintain")
    side: str | None = None


class PeerDistribution(BaseModel):
    """연령대 또래 lifestyle_score 분포 히스토그램 (리포트 또래 곡선용)."""

    counts: list[int]
    edges: list[float]
    my_bin: int


class LifestyleShap(BaseModel):
    items: list[ShapItem]
    lifestyle_score: float
    peer_top_pct: int | None = None
    peer_relative: str | None = None
    peer_distribution: PeerDistribution | None = None


class ClinicalItem(BaseModel):
    """모델1 임상 검진 항목 상세 (리포트 임상 상세표)."""

    feature: str
    label: str
    desc: str
    category: str
    normal_range: str
    value_text: str
    status: str  # 상태 라벨 (정상/주의/위험/낮음/높음/복부비만 …)
    status_level: str  # good/info/caution/warnLight/danger
    direction: str  # low/high/normal  (정상범위 대비 — 프론트가 ▽/🔺 렌더)
    disease_low: str  # 미달 시 의심 원인
    disease_high: str  # 초과 시 위험


class LifestyleItem(BaseModel):
    """모델2 생활습관 항목 상세 (리포트 생활습관 상세표)."""

    feature: str
    label: str
    normal_range: str
    value_text: str
    status: str
    status_level: str
    group: str  # improve | maintain
    action: str = ""  # 개선 시 행동 제안 (improve 항목)
    domain: str = ""  # diet | exercise | etc (Phase B)


class LifestyleDomainSummary(BaseModel):
    """생활습관 도메인별 핵심요약 (Phase B). 항상 식이/운동/기타 3개."""

    domain: str  # diet | exercise | etc
    domain_label: str  # 식이 | 운동 | 기타
    improve_count: int  # 해당 도메인 개선 필요 항목 수
    summary: str  # 규칙 기반 한 줄


class ReportMeta(BaseModel):
    """리포트 메타 (그룹·점수·인구통계·기저질환·또래비교)."""

    group: str | None  # app_group (A~D) ← 서비스 내부 G1~G4 코드 그대로
    group_title: str  # 신장 집중 관리군 …
    grade: str  # 높음/주의/낮음
    score: float | None  # CKD 위험도 선별 점수 (0~100, = ckd_risk_score*100)
    group_message: str
    age: int | None
    gender: str | None  # "남성"/"여성"
    conditions: list[str]  # 진단 기저질환 (고혈압/당뇨/이상지질혈증/CKD)
    family_history: list[str]  # 가족력 (고혈압/당뇨/심장질환)
    peer_top_pct: int | None
    peer_relative: str | None
    report_available: bool = True  # 진단자(CKD/DIALYSIS)면 False — 위험도 예측·리포트 비대상


class ReportResponse(BaseSerializerModel):
    health_check_id: int
    shap_model1: list[ShapItem]
    shap_model2: LifestyleShap | None
    ai_guide: str
    recommended_tests: list[str] = []  # 모델1(app_group) 기반 권장 검사
    model1_summary: str = ""  # 모델1 종합 한 줄 요약
    # ── A2: 임상·생활습관 상세표 + 리포트 메타 (순수 추가, 기존 필드 불변) ──
    clinical_items: list[ClinicalItem] = []
    lifestyle_items: list[LifestyleItem] = []
    lifestyle_domain_summary: list[LifestyleDomainSummary] = []  # Phase B: 도메인별 요약
    report_meta: ReportMeta | None = None

File: frontend/ckd-care-app/src/api/healthCheck.ts
import { api, BASE, getToken } from "./client";

export type DialysisType = "none" | "hemodialysis" | "peritoneal" | "transplant";
export type UrineResult = "POSITIVE" | "NEGATIVE";

export interface HealthCheckCreateRequest {
  checked_date: string;
  systolic_bp: number;
  diastolic_bp: number;
  fasting_glucose: number;
  creatinine?: number | null;
  total_cholesterol?: number | null;
  hdl_cholesterol?: number | null;
  triglycerides?: number | null;
  weight: number;
  height: number;
  waist_circumference?: number | null;
  // 신규 검진 항목 (ML 모델 입력에 포함)
  ldl_cholesterol?: number | null;
  hemoglobin?: number | null;
  ast?: number | null;
  alt?: number | null;
  urine_protein?: UrineResult | null;
  urine_glucose?: UrineResult | null;
  // dialysis_type 제거 — CKD 진단·투석 종류는 문진(LifestyleSurvey)이 단일 진실
}

export interface HealthCheckResponse {
  id: number;
  user_id: number;
  checked_date: string;
  systolic_bp: number;
  diastolic_bp: number;
  fasting_glucose: number;
  creatinine: number | null;
  total_cholesterol: number | null;
  hdl_cholesterol: number | null;
  triglycerides: number | null;
  weight: number;
  height: number;
  bmi: number;
  waist_circumference: number | null;
  ldl_cholesterol: number | null;
  hemoglobin: number | null;
  ast: number | null;
  alt: number | null;
  urine_protein: UrineResult | null;
  urine_glucose: UrineResult | null;
  egfr_estimated: number | null;
  ckd_risk_score: number | null;
  ckd_stage: string | null;
  safety_warning: string | null;
  dialysis_type: DialysisType | null;
  created_at: string;
}

export interface HealthCheckListResponse {
  total: number;
  items: HealthCheckResponse[];
}

// ===== SHAP 리포트 타입 =====

export interface ShapItem1 {
  feature: string;
  value: number;
  shap: number;
  note?: string;
  side?: "improve" | "maintain" | "exclude";
}

export interface LifestyleShapItem {
  feature: string;
  value: number;
  shap: number;
  side?: "improve" | "maintain";
}

export interface PeerDistribution {
  counts: number[];
  edges: number[];
  my_bin: number;
}

export interface LifestyleShap {
  items: LifestyleShapItem[];
  lifestyle_score: number;
  peer_top_pct: number | null;
  peer_relative: string | null;
  peer_distribution?: PeerDistribution | null;
}

// ===== 임상 상세 분석 타입 (Phase A 백엔드 추가분) =====

export interface ClinicalItem {
  feature: string;
  label: string;
  desc: string;
  category: string;
  normal_range: string;
  value_text: string;
  status: string;
  status_level: "good" | "info" | "caution" | "warnLight" | "danger";
  direction: "low" | "high" | "normal";
  disease_low: string;
  disease_high: string;
}

export interface LifestyleItem {
  feature: string;
  label: string;
  normal_range: string;
  value_text: string;
  status: string;
  status_level: "good" | "info" | "caution" | "warnLight" | "danger";
  group: "improve" | "maintain";
  action: string;
  domain: string; // diet | exercise | etc (Phase B)
}

// ===== 생활습관 도메인 분리 타입 (Phase B) =====

export interface LifestyleDomainSummary {
  domain: string;
  domain_label: string;
  improve_count: number;
  summary: string;
}

export interface ReportMeta {
  group: string | null;
  group_title: string;
  grade: string;
  score: number | null;
  group_message: string;
  age: number | null;
  gender: string | null;
  conditions: string[];
  family_history: string[];
  peer_top_pct: number | null;
  peer_relative: string | null;
}

export interface ReportResponse {
  health_check_id: number;
  shap_model1: ShapItem1[];
  shap_model2: LifestyleShap | null;
  ai_guide: string;
  recommended_tests?: string[];
  model1_summary?: string;
  clinical_items?: ClinicalItem[];
  lifestyle_items?: LifestyleItem[];
  lifestyle_domain_summary?: LifestyleDomainSummary[]; // 도메인별 생활습관 요약 (Phase B)
  report_meta?: ReportMeta | null;
}

// ===== API =====

// ===== OCR 응답 타입 =====
export interface OCRField {
  text: string;
  confidence: number; // 0~1
}

export interface OCRMappedField {
  value: number;
  confidence: number;
  source_text: string;
}

// 자동 매핑되는 검진 필드 (ManualInputPage prefill용 필드명과 일치)
// LDL은 시스템에서 사용 안 함 → OCR 매핑에서 제외
export type OCRMappedKey =
  | "fasting_glucose"
  | "creatinine"
  | "hdl_cholesterol"
  | "total_cholesterol"
  | "triglycerides"
  | "systolic_bp"
  | "diastolic_bp"
  | "height"
  | "weight"
  | "waist_circumference";

export interface OCRLine {
  text: string;
  confidence: number;
}

export interface OCRResponse {
  engine: "clova" | "stub";
  filename: string;
  fields: OCRField[];
  lines?: OCRLine[];
  mapped?: Partial<Record<OCRMappedKey, OCRMappedField>>;
  low_confidence_count: number;
  page_count?: number;
  page_errors?: string[];
}

export const healthCheckApi = {
  create: (body: HealthCheckCreateRequest) =>
    api.post<HealthCheckResponse>("/health-checks", body),
  list: (limit = 20, offset = 0) =>
    api.get<HealthCheckListResponse>(`/health-checks?limit=${limit}&offset=${offset}`),
  getReport: (id: number) =>
    api.get<ReportResponse>(`/health-checks/${id}/report`),
  delete: (id: number) =>
    api.delete<void>(`/health-checks/${id}`),
  update: (id: number, body: HealthCheckCreateRequest) =>
    api.patch<HealthCheckResponse>(`/health-checks/${id}`, body),
  deleteAll: () =>
    api.delete<{ deleted_count: number }>("/health-checks"),
  ocrExtract: (file: File) => {
    const fd = new FormData();
    fd.append("file", file);
    return api.postForm<OCRResponse>("/health-checks/ocr", fd);
  },
  downloadPdf: async (id: number): Promise<void> => {
    const token = getToken();
    const res = await fetch(`${BASE}/health-checks/${id}/pdf`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      credentials: "include",
    });
    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new Error(`PDF 다운로드 실패 (${res.status}): ${body.slice(0, 200)}`);
    }
    const blob = await res.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `건강리포트_${new Date().toISOString().slice(0, 10)}.pdf`;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
  },
};
